pkg/adaptor/etcd/writer.go

```diff
@@ -0,0 +1,100 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/ops"
+
+	eclient "github.com/coreos/etcd/client"
+)
+
+var (
+	_ client.Writer = &Writer{}
+)
+
+// Writer implements client.Writer for use with etcd
+type Writer struct {
+	rootKey  string
+	writeMap map[ops.Op]func(message.Msg, eclient.KeysAPI) error
+}
+
+func newWriter(rootKey string) *Writer {
+	w := &Writer{
+		rootKey: rootKey,
+	}
+	w.writeMap = map[ops.Op]func(message.Msg, eclient.KeysAPI) error{
+		ops.Insert: w.insertMsg,
+		ops.Update: w.updateMsg,
+		ops.Delete: w.deleteMsg,
+	}
+	return w
+}
+
+func (w *Writer) Write(msg message.Msg) func(client.Session) error {
+	return func(s client.Session) error {
+		w, ok := w.writeMap[msg.OP()]
+		if !ok {
+			log.Infof("no function registered for operation, %s\n", msg.OP())
+			return nil
+		}
+		return w(msg, keysAPI(s.(*Session)))
+	}
+}
+
+func keysAPI(s *Session) eclient.KeysAPI {
+	return eclient.NewKeysAPI(s.Client)
+}
+
+func (w *Writer) genKey(m message.Msg, dataKey string) (string, error) {
+	_, ns, err := message.SplitNamespace(m)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("/%s/%s/%s", w.rootKey, ns, dataKey), nil
```

Comment:
The `reader` creates a namespace by replacing all `/` with `_`. If we use transporter to copy an etcd into another etcd, ideally we would want the data to be equivalent. I don't think that this is the case in this implementation... I think the keys would all end up under one dir with the `_` replacement

Maybe the reader should preserve `/`? What was the reasoning behind the `_` replacement? It is difficult to map the tree structure of etcd to our message format because we only consider one level of namespacing, while etcd can have nested keys. Not sure what we want to do here... maybe it is okay not to have A->B->A idempotence in ETCD.

Depending on this decision, we might have to tweak the insert function to create proper nested dirs/keys.